File: src/types/assembly.ts
// === アセンブリ管理型定義 ===
// プログラムをコアアセンブリとアプリケーションアセンブリに分離管理する

/** ビルドターゲットプラットフォーム */
export type BuildTarget = 'webgl' | 'pc';

/** ビルド方式 */
export type BuildMethod = 'typescript-jit' | 'wasm-bundle';

/** ビルドターゲットとビルド方式の対応 */
export interface BuildConfig {
  target: BuildTarget;
  method: BuildMethod;
}

/** プラットフォームごとのデフォルトビルド方式 */
export const DEFAULT_BUILD_CONFIGS: BuildConfig[] = [
  { target: 'webgl', method: 'typescript-jit' },  // WebGL: TypeScript JIT (ホットリロード対応)
  { target: 'pc', method: 'typescript-jit' },      // PC: TypeScript JIT (Tauri webview内実行)
];

// ─────────────────────────────────────────────
// コアアセンブリ
// PictorやErgoから取得した変更頻度が低い使いまわせるコード
// ─────────────────────────────────────────────

/** コアアセンブリの取得元 */
export type CoreAssemblyOrigin = 'pictor' | 'ergo' | 'other';

/** コアアセンブリのアーティファクト種別 */
export type CoreArtifactType = 'source' | 'wasm';

/** コアアセンブリのカスタマイズ状態 */
export type CustomizationStatus = 'original' | 'modified';

/** コアアセンブリ定義 */
export interface CoreAssembly {
  /** 一意識別子 */
  id: string;
  /** 表示名 */
  name: string;
  /** 取得元 (Pictor, Ergo, etc.) */
  origin: CoreAssemblyOrigin;
  /** リリースデポ上のパス/識別子 */
  depotPath: string;
  /** リリースデポのバージョンタグ */
  version: string;
  /** アーティファクト種別 (ソースコード or DLL) */
  artifactType: CoreArtifactType;
  /** カスタマイズ状態 */
  customization: CustomizationStatus;
  /** 方言対応のカスタマイズメモ (modified時のみ) */
  customizationNote?: string;
  /** プロジェクトローカルでのパス */
  localPath?: string;
  /** 対応ビルドターゲット */
  buildTargets: BuildTarget[];
}

// ─────────────────────────────────────────────
// アプリケーションアセンブリ
// ゲーム固有のプログラム。モノリシックで使いまわせないもの
// アプリケーションやシーン別にアセンブリを作成して分散的にビルド
// ─────────────────────────────────────────────

/** アプリケーションアセンブリのスコープ */
export type AssemblyScope = 'application' | 'scene';

/** アプリケーションアセンブリ定義 */
export interface ApplicationAssembly {
  /** 一意識別子 */
  id: string;
  /** 表示名 */
  name: string;
  /** スコープ (アプリケーション全体 or シーン固有) */
  scope: AssemblyScope;
  /** シーンスコープの場合、紐づくシーンID */
  sceneId?: string;
  /** ソースコードのローカルパス */
  sourcePath: string;
  /** 各ターゲットごとのビルド設定 */
  buildConfigs: BuildConfig[];
  /** 依存するコアアセンブリID一覧 */
  coreAssemblyDependencies: string[];
  /** 依存する他のアプリケーションアセンブリID一覧 */
  appAssemblyDependencies: string[];
  /** エントリポイントファイル名 */
  entryPoint?: string;
}

// ─────────────────────────────────────────────
// プロジェクトレベルのアセンブリ管理設定
// ─────────────────────────────────────────────

/** リリースデポ接続設定 */
export interface ReleaseDepotConfig {
  /** デポの名前 */
  name: string;
  /** デポのURL/パス */
  url: string;
  /** 認証トークン (オプション) */
  authToken?: string;
}

/** リソースデポ参照設定 (実体は別システム) */
export interface ResourceDepotRef {
  /** デポの名前 */
  name: string;
  /** デポのURL/パス */
  url: string;
}

/** データオーガナイザー参照設定 (実体は別システム) */
export interface DataOrganizerRef {
  /** オーガナイザーの名前 */
  name: string;
  /** オーガナイザーのURL/パス */
  url: string;
}

/** プロジェクトのアセンブリ管理設定 */
export interface ProjectAssemblyConfig {
  /** リリースデポ設定一覧 */
  releaseDepots: ReleaseDepotConfig[];
  /** コアアセンブリ一覧 */
  coreAssemblies: CoreAssembly[];
  /** アプリケーションアセンブリ一覧 */
  applicationAssemblies: ApplicationAssembly[];
  /** リソースデポ参照 (外部システム) */
  resourceDepotRef?: ResourceDepotRef;
  /** データオーガナイザー参照 (外部システム) */
  dataOrganizerRef?: DataOrganizerRef;
}

/** デフォルトのプロジェクトアセンブリ設定 */
export function createDefaultAssemblyConfig(): ProjectAssemblyConfig {
  return {
    releaseDepots: [],
    coreAssemblies: [],
    applicationAssemblies: [],
  };
}

File: src-tauri/src/models/assembly.rs
use serde::{Deserialize, Serialize};

/// バックエンドプラットフォーム
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum BackendPlatform {
    /// Ars独自ランタイム (TypeScript JIT / WASM)
    ArsNative,
    /// Unity (C#)
    Unity,
    /// Unreal Engine (C++)
    Unreal,
    /// Godot (GDScript)
    Godot,
}

impl Default for BackendPlatform {
    fn default() -> Self {
        Self::ArsNative
    }
}

/// バックエンドプラットフォーム設定
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendPlatformConfig {
    /// 選択中のバックエンドプラットフォーム
    pub platform: BackendPlatform,
    /// プラットフォーム固有の設定
    #[serde(skip_serializing_if = "Option::is_none")]
    pub platform_options: Option<serde_json::Value>,
}

impl Default for BackendPlatformConfig {
    fn default() -> Self {
        Self {
            platform: BackendPlatform::default(),
            platform_options: None,
        }
    }
}

/// ビルドターゲットプラットフォーム
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum BuildTarget {
    Webgl,
    Pc,
}

/// ビルド方式
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum BuildMethod {
    /// TypeScript JIT: ホットリロード対応、WebGL/PC共通
    TypescriptJit,
    /// WASM Bundle: Emscriptenコンパイル済みC++コアエンジン
    WasmBundle,
}

/// ビルド設定
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BuildConfig {
    pub target: BuildTarget,
    pub method: BuildMethod,
}

/// コアアセンブリの取得元
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum CoreAssemblyOrigin {
    Pictor,
    Ergo,
    Other,
}

/// コアアセンブリのアーティファクト種別
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum CoreArtifactType {
    /// TypeScriptソースコード
    Source,
    /// Emscriptenコンパイル済みWASMバイナリ
    Wasm,
}

/// カスタマイズ状態
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum CustomizationStatus {
    Original,
    Modified,
}

/// コアアセンブリ定義
/// PictorやErgoから取得した変更頻度が低い使いまわせるコード
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CoreAssembly {
    pub id: String,
    pub name: String,
    pub origin: CoreAssemblyOrigin,
    pub depot_path: String,
    pub version: String,
    pub artifact_type: CoreArtifactType,
    pub customization: CustomizationStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub customization_note: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub local_path: Option<String>,
    pub build_targets: Vec<BuildTarget>,
}

/// アプリケーションアセンブリのスコープ
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum AssemblyScope {
    Application,
    Scene,
}

/// アプリケーションアセンブリ定義
/// ゲーム固有のプログラム。モノリシックで使いまわせないもの
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApplicationAssembly {
    pub id: String,
    pub name: String,
    pub scope: AssemblyScope,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scene_id: Option<String>,
    pub source_path: String,
    pub build_configs: Vec<BuildConfig>,
    pub core_assembly_dependencies: Vec<String>,
    pub app_assembly_dependencies: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entry_point: Option<String>,
}

/// リリースデポ接続設定
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReleaseDepotConfig {
    pub name: String,
    pub url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth_token: Option<String>,
}

/// リソースデポ参照設定 (実体は別システム)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceDepotRef {
    pub name: String,
    pub url: String,
}

/// データオーガナイザー参照設定 (実体は別システム)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataOrganizerRef {
    pub name: String,
    pub url: String,
}

/// プロジェクトのアセンブリ管理設定
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectAssemblyConfig {
    /// バックエンドプラットフォーム設定
    #[serde(default)]
    pub backend_platform: BackendPlatformConfig,
    pub release_depots: Vec<ReleaseDepotConfig>,
    pub core_assemblies: Vec<CoreAssembly>,
    pub application_assemblies: Vec<ApplicationAssembly>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resource_depot_ref: Option<ResourceDepotRef>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data_organizer_ref: Option<DataOrganizerRef>,
}

impl Default for ProjectAssemblyConfig {
    fn default() -> Self {
        Self {
            backend_platform: BackendPlatformConfig::default(),
            release_depots: Vec::new(),
            core_assemblies: Vec::new(),
            application_assemblies: Vec::new(),
            resource_depot_ref: None,
            data_organizer_ref: None,
        }
    }
}
